### PR TITLE
Added user authentication and authorization

### PR DESCRIPTION
**Pull request by katief906**
Added user authentication and authorization methods to our HomesController and our API::V1::ShopsController
Updated header to only show "Add New Shop" to users that are admins
Require users to login before they can add a new shop (and they have to be admins to do so)
Still need to write controller and feature tests

### CODE REVIEW

File: app/controllers/api/v1/shops_controller.rb [hunk 1]
```diff
@@ -1,4 +1,7 @@
 class Api::V1::ShopsController < ApiController 
+  before_action :authenticate_user, except: [:index, :show]
+  before_action :authorize_user, except: [:index, :show, :create]
```
Comment (dovakeen118) [inline app/controllers/api/v1/shops_controller.rb]: we don't need to duplicate these methods to be used in both index and show - both methods check `if !user_signed_in?`, just the `authorize` also checks if the current user is an admin. Looks like `authorize` should be used for `create`, then index and show is just concerned with a user being signed in, thus using `authenticate`

File: app/controllers/application_controller.rb [hunk 2]
```diff
@@ -1,5 +1,13 @@
 class ApplicationController < ActionController::Base
   # Prevent CSRF attacks by raising an exception.
   # For APIs, you may want to use :null_session instead.
-  protect_from_forgery with: :exception
+  protect_from_forgery unless: -> { request.format.json? }
```
Comment (dovakeen118) [inline app/controllers/application_controller.rb]: this regular ApplicationController to be used by requests with Rails view templates should keep the existing protect from forgery line. This updated version with `request.format.json?` belongs in the ApiController as it's only needed when making API requests (which you already have)

File: app/controllers/homes_controller.rb [hunk 3]
```diff
@@ -1,4 +1,22 @@
 class HomesController < ApplicationController
+  before_action :authenticate_user!, except: [:index, :authorized]
+  before_action :authorize_user, except: [:index, :authenticated]
```
Comment (dovakeen118) [inline app/controllers/homes_controller.rb]: same idea here - both methods check if a user is signed in, so we don't need to repeat actions between the 2. `authenticate_user!` just checks if a user is signed in
`authorize_user` _also_ checks if they are an admin

for `index`, looks like you want just `authenticate_user!` to check if they're signed in, not caring if they are an admin

File: db/migrate/20221031193754_devise_create_users.rb [hunk 4]
```diff
@@ -4,6 +4,9 @@ class DeviseCreateUsers < ActiveRecord::Migration[5.2]
   def change
     create_table :users do |t|
       ## Database authenticatable
+      t.string :first_name, null: false
+      t.string :last_name, null: false
+      t.string :role, null: false, default: "member"
```
Comment (dovakeen118) [inline db/migrate/20221031193754_devise_create_users.rb:9]: Not something that needs to be changed right now, but generally when setting up a `role` we're better off using an `enum` data type than just a string. Having a string data type means we could have users with roles of "member", "admin", "cat", "mmber" (spelt incorrectly) leaving the implementation fragile. Whereas an enum allows us to designate the options that are allowed!

**Review by dovakeen118 — COMMENTED** [review]
a couple comments to clean up the `before_action` hooks in your controllers!

**Comment by katief906** [comment]
Kerrin, I responded to all of your comments minus the last one about the enum data type since we need some more time to review that documentation

**Pull request merged**